File: currying/currying2.js
'use strict';

var people = [
  { name: 'Calvin' },
  { name: 'John' },
  { name: 'Thomas' }
];

// function with hardcoded key 'name'
function getPersonName(obj) {
  return obj['name'];
}

// mapping to the hardcoded function
var names = people.map(getPersonName);
console.log(names);

// Specify the key on-the-fly by using a generic function.
function getByKey(key, obj) {
  return function(obj) {
    return obj[key];
  };
}

var getByKeyPartial = getByKey('name');
console.log(getByKeyPartial);

var names2 = people.map(getByKeyPartial);
console.log(names2);

var names3 = people.map(getByKey('name'));
console.log(names3);

// mapping to anonymous function that uses generic get
//var names2 = people.map(function(obj) { 
  //return getByKey('name', obj);
//});
//console.log(names2);

//var getByKeyPartial = getByKey('name');
//console.log(getByKeyPartial);

//// We can nest our generic get function further to achieve a simpler interface
//function get(key) {
  //return people.map(function(obj){
    //return getByKey(key, obj);
  //});
//}

//var names3 = get('name');
//console.log(names3);

//// the people array is still hardcoded so we can make it even more generic, like this
//function g(key, arr) {
  //return arr.map(function(obj) {
    //return getByKey(key, obj);
  //});
//}

//var names4 = g('name', people);
//console.log(names4);

File: currying/currying2.hs
module Main where

people ::  [(String, String)]
people = [("name", "Calvin")
    ,("wat", "John")
    ,("name", "Thomas")
    ]

{- generic filterByKey function that accepts a string as keyname -}
filterByKey :: Eq a => a -> [(a, t)] -> [t]
filterByKey _ [] = []
filterByKey p ((k, v):xs)
    | p == k    = v : filterByKey p xs
    | otherwise = filterByKey p xs

filterByName :: [(String, t)] -> [t]
filterByName = filterByKey "name"

getName :: [(a, b)] -> [b]
getName = map snd

main ::  IO ()
main = do
    let names = getName people  {- snd simply plucks out the 2nd value in the tuple -}
    print names
    let names2 = filterByKey "name" people
    print names2
    let names3 = filterByName people
    print names3
